RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc

```diff
@@ -0,0 +1,198 @@
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+
+#include "PFRecHitProducerKernel.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  using namespace ParticleFlowRecHitProducer;
+
+  // Kernel to apply cuts to calorimeter hits and construct PFRecHits
+  template <typename CAL>
+  struct PFRecHitProducerKernelConstruct {
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const typename CAL::ParameterType::ConstView params,
+                                  const typename CAL::CaloRecHitSoATypeDevice::ConstView recHits,
+                                  reco::PFRecHitDeviceCollection::View pfRecHits,
+                                  uint32_t* __restrict__ denseId2pfRecHit,
+                                  uint32_t* __restrict__ num_pfRecHits) const {
+      // Strided loop over CaloRecHits
+      for (int32_t i : cms::alpakatools::elements_with_stride(acc, recHits.metadata().size())) {
+        // Check energy thresholds/quality cuts (specialised for HCAL/ECAL)
+        if (!applyCuts(recHits[i], params))
+          continue;
+
+        // Use atomic operation to determine index of the PFRecHit to be constructed
+        // The index needs to be unique and consequtive across all threads in all blocks.
+        // This is achieved using the alpaka::hierarchy::Blocks argument.
+        const uint32_t j = alpaka::atomicAdd(acc, num_pfRecHits, 1u, alpaka::hierarchy::Blocks{});
```

Comment:
Can you check if `alpaka::atomicInc` works correctly here ?
```suggestion
        const uint32_t j = alpaka::atomicInc(acc, num_pfRecHits, 0xffffffff, alpaka::hierarchy::Blocks{});
```

In principle it should generate more efficient code, at least for the CUDA backend.

For the CPU backend is shouldn't matter anyway, since operations with the `alpaka::hierarchy::Blocks{}` argument should decay to the standard, non-atomic version for the serial backend.

Reply:
It works, but there is no measurable performance difference on the CUDA backend. I will include this in the next commit.